src/libre/ast_analysis.c

```diff
@@ -153,77 +252,326 @@ analysis_iter(struct ast_expr *n)
+		    __func__, ast_node_type_name(n->type));
+		return UA_YES;
+
+	case AST_EXPR_CONCAT: {
+		size_t i;
+
+		/* has unanchored end if last non-nullable child does */
+		for (i = n->u.concat.count; i > 0; i--) {
+			const struct ast_expr *child = n->u.concat.n[i - 1];
+			const enum unanchored_analysis_res child_res = analysis_iter_unanchored_end(child);
+			if (child_res != UA_UNDECIDED) {
+				LOG(3 - LOG_UNANCHORED_FLAGS, "%s: %s -> %c (child res)\n",
+				    __func__, ast_node_type_name(n->type), child_res);
+				return child_res;
+			}
+
+			if (!is_nullable(child)) {
+				break;
+			}
+		}
+
+		break;
+	}
+
+	case AST_EXPR_ALT: {
+		size_t i;
+
+		/* if all children agree, return that result, otherwise undecided */
+		const enum unanchored_analysis_res first_child_res = analysis_iter_unanchored_end(n->u.alt.n[0]);
+		LOG(3 - LOG_UNANCHORED_FLAGS, "%s: ALT child 0 -- %s -> %c (child res)\n",
+		    __func__, ast_node_type_name(n->type), first_child_res);
+
+		for (i = 1; i < n->u.alt.count; i++) {
+			const struct ast_expr *child = n->u.alt.n[i];
+			const enum unanchored_analysis_res child_res = analysis_iter_unanchored_end(child);
+			LOG(3 - LOG_UNANCHORED_FLAGS, "%s: ALT child %zd -- %s -> %c (child res)\n",
+			    __func__, i, ast_node_type_name(n->type), child_res);
+			if (child_res != first_child_res) {
+				LOG(3 - LOG_UNANCHORED_FLAGS, "%s: %s -> %c (child result) contracts first, returning U\n",
+				    __func__, ast_node_type_name(n->type), child_res);
+				return UA_UNDECIDED;
+			}
+		}
+
+		LOG(3 - LOG_UNANCHORED_FLAGS, "%s: %s -> %c (all children agree)\n",
+		    __func__, ast_node_type_name(n->type), first_child_res);
+		return first_child_res;
+	}
+
+	case AST_EXPR_REPEAT:
+		if (n->u.repeat.min == 0) {
+			LOG(3 - LOG_UNANCHORED_FLAGS, "%s: %s -> U (repeat.min == 0)\n",
+			    __func__, ast_node_type_name(n->type));
+			return UA_UNDECIDED;
+		}
+		return analysis_iter_unanchored_end(n->u.repeat.e);
 
 	case AST_EXPR_GROUP:
-		return always_consumes_input(n->u.group.e);
+		return analysis_iter_unanchored_end(n->u.group.e);
+
+	case AST_EXPR_SUBTRACT:
+		return analysis_iter_unanchored_end(n->u.subtract.a);
 
 	default:
-		return 0;
+		assert(!"unreached");
+	}
+
+	return UA_UNDECIDED;
+}
+
+static void
+set_flags_subtree(struct ast_expr *n, enum ast_flags flags)
+{
+	n->flags |= flags;
+
+	switch (n->type) {
+	case AST_EXPR_EMPTY:
+	case AST_EXPR_TOMBSTONE:
+	case AST_EXPR_ANCHOR:
+	case AST_EXPR_LITERAL:
+	case AST_EXPR_CODEPOINT:
+	case AST_EXPR_RANGE:
+		break;
+
+	case AST_EXPR_CONCAT: {
+		for (size_t i = 0; i < n->u.concat.count; i++) {
+			struct ast_expr *child = n->u.concat.n[i];
+			set_flags_subtree(child, flags);
+		}
+		break;
+	}
+
+	case AST_EXPR_ALT: {
+		for (size_t i = 0; i < n->u.alt.count; i++) {
+			struct ast_expr *child = n->u.alt.n[i];
+			set_flags_subtree(child, flags);
+		}
+		break;
+	}
+
+	case AST_EXPR_REPEAT:
+		set_flags_subtree(n->u.repeat.e, flags);
```

Comment:
Is the fall-through here intentional? It doesn't look like it! If it's not, I'm surprised fuzzing didn't catch this.

Reply:
Oh, yikes, good catch. Those happen to be safe though coincidence, because all three fields in the union occur at the same offset and repeatedly setting the same flags has no effect, but that's definitely a bug. Fixed in 4cfe6e34.

Reply:
oh! that's why it didn't show. cool!